cumulusci/tasks/salesforce/composite.py

```diff
@@ -0,0 +1,120 @@
+
+from cumulusci.cli.ui import CliTable
+from cumulusci.core.exceptions import SalesforceException
+from cumulusci.core.utils import process_bool_arg
+from cumulusci.tasks.salesforce import BaseSalesforceApiTask
+from cumulusci.utils import inject_namespace
+
+API_ROLLBACK_MESSAGE = "The transaction was rolled back since another operation in the same transaction failed."
+API_INVALID_REF_MESSAGE = "Invalid reference specified."
+
+
+class CompositeApi(BaseSalesforceApiTask):
+    """
+    Wrapper for Composite REST API calls.
+
+    Replaces "%%%USERID%%%" and namespace tokens with the running user's
+    ID and project namespace, respectively. Replaces any usernames that include
+    the "connected.edu" domain with a random TLD.
+    """
+
+    task_options = {
+        "data_files": {
+            "description": "A list of paths to the json files to POST.",
+            "required": True,
+        },
+        "managed": {
+            "description": "If True, replaces namespace tokens with the namespace prefix.",
+            "default": False,
+        },
+        "namespaced": {
+            "description": "If True, replaces namespace tokens with the namespace prefix.",
+            "default": False,
+        },
+        "randomize_username": {
+            "description": "If True, randomize the TLD for any 'Username' fields.",
+            "default": False,
+        },
+    }
+
+    def _run_task(self):
+        for data_file_path in self.options["data_files"]:
+            self.logger.info(f"Processing {data_file_path}")
+            self._composite_request(data_file_path)
+
+    def _composite_request(self, data_file_path):
+        request_body = self._process_json(Path(data_file_path).read_text())
+        self.is_all_or_none = json.loads(request_body)["allOrNone"]
+        client = self._init_api()
+        result = client.restful("composite", method="POST", data=request_body)
+        self._process_response(result)
+
+    def _process_json(self, body):
+        """Replace namespace tokens and randomize username domains."""
+        user_id = self.org_config.user_id
+        body = body.replace("%%%USERID%%%", user_id)
+
+        namespace = self.project_config.project__package__namespace
+        if "managed" in self.options:
+            managed = process_bool_arg(self.options["managed"])
+        else:
+            managed = (
+                bool(namespace) and namespace in self.org_config.installed_packages
+            )
+
+        _, body = inject_namespace(
+            "composite",
+            body,
+            namespace=namespace,
+            managed=managed,
+            namespaced_org=self.options.get("namespaced", self.org_config.namespaced),
+        )
+
+        if self.options.get("randomize_username", False):
+            random_tld = "".join(random.choices(string.ascii_lowercase, k=4))
+            body = re.sub(
+                r'("Username": .[\w-]+@[\w-]+\.)+[\w-]+', fr"\1{random_tld}", body
+            )
```

Comment:
I didn't think it a significant risk since `User.Email` is distinct. The goal here is avoid username collisions by modifying the value just enough to be unique but recognizable.

Reply:
`User.Email` is distinct? I don't think it is.

And in any case the situation I'm worried about would apply even if there is no other Salesforce user with the same email. I'm concerned that we might randomly generate a real 4-letter TLD and then send email to someone with no connection to the org, even if that's not super likely.

Reply:
Since `User.Email` is used for system generated emails instead of `User.Username` I don't think the situation you describe is possible.

Reply:
@jstvz doh, yeah, I had it turned around in my head even though it was allegedly the thing I was trying to pay attention to. Thanks.